File: tests/src/bench_bulk_insert.rs
use std::cell::{Cell, RefCell};
use std::io;
use std::io::Write;
use std::rc::Rc;

use serde::{Deserialize, Serialize};
use tester::{Bencher, TDynBenchFn};

use tarantool::fiber::Fiber;
use tarantool::net_box::{Conn, ConnOptions, Options};
use tarantool::tuple::Encode;

pub struct BulkInsertBenchmark {
    pub test_size: usize,
    pub num_fibers: usize,
    pub num_rows: usize,
}

#[derive(Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct S1Record<'a> {
    pub id: u32,
    pub text: &'a str,
}

impl Encode for S1Record<'_> {}

impl TDynBenchFn for BulkInsertBenchmark {
    fn run(&self, harness: &mut Bencher) {
        let test_size = self.test_size;
        let num_fibers = self.num_fibers;
        let num_rows = self.num_rows;

        let text = RefCell::new("X".repeat(test_size));
        let id_counter: Cell<usize> = Cell::new(0);

        let conn = Rc::new(
            Conn::new(
                ("localhost", unsafe { crate::LISTEN }),
                ConnOptions {
                    user: "bench_user".to_string(),
                    password: "password".to_string(),
                    ..ConnOptions::default()
                },
                None,
            )
            .unwrap(),
        );

        conn.call("_cleanup", Vec::<()>::new().as_slice(), &Options::default())
            .unwrap();

        harness.iter(|| {
            let mut fiber_pool = Vec::with_capacity(num_fibers);
            for _ in 0..num_fibers {
                let mut fiber = Fiber::new("bench_fiber", &mut |base_id| {
                    let base_id = *base_id;
                    let mut row = S1Record {
                        id: base_id as u32,
                        text: &*text.borrow(),
                    };

                    let mut remote_space = conn.space("bench_s1").unwrap().unwrap();
                    let insert_options = Options {
                        ..Options::default()
                    };

                    for _ in 0..num_rows {
                        let insert_result = remote_space.insert(&row, &insert_options).unwrap();
                        assert!(insert_result.is_some());
                        row.id += 1;
                    }
                    0
                });
                fiber.set_joinable(true);
                fiber_pool.push(fiber)
            }

            for fiber in fiber_pool.iter_mut() {
                let id = id_counter.get();
                id_counter.set(id + num_rows);
                fiber.start(id)
            }

            for (_, fiber) in fiber_pool.iter().enumerate() {
                fiber.join();
            }

            print!(".");
            io::stdout().flush().unwrap();
        });
    }
}

File: tests/src/bench_bulk_insert.lua
#!/usr/bin/env tarantool

box.cfg{
    listen = 3302,
    wal_mode = 'none'
}

------------------------------------------------------------------------------

clock = require('clock')
fiber = require('fiber')
log = require('log')
net_box = require('net.box')

local test_size = 64;
local num_fibers = 256;
local num_rows = 1000;
local num_passes = 30;

local text = string.rep('X', test_size)
local pass_times = {}

------------------------------------------------------------------------------

local conn = net_box.connect('bench_user:password@127.0.0.1:3301')
conn:call('_cleanup')

local id_base = 1
for pass_id = 1, num_passes do
    local begin_time = tonumber(clock.monotonic64())
    local local_space = box.space.bench_s1

    local fiber_pool = {}
    for fiber_id = 1, num_fibers do
        local fiber = fiber.create(
            function(id_base)
                local remote_space = conn.space.bench_s1
                for id = id_base, (id_base + num_rows - 1)  do
                    local res = pcall(remote_space.insert, remote_space, {id + id_base, text})
                    if not res then
                        io.write('x')
                    end
                end
            end,
            id_base
        )
        fiber:set_joinable(true)
        table.insert(fiber_pool, fiber)
        id_base = id_base + num_rows
    end

    for i = 1, num_fibers do
        local fiber = fiber_pool[i]
        if fiber:status() ~= 'dead' then
            fiber:join()
        end
    end

    local end_time = tonumber(clock.monotonic64())
    table.insert(pass_times, end_time - begin_time)
    io.write('.')
    io.flush()
end

conn:close()

------------------------------------------------------------------------------

local avg = 0
for i = 1, num_passes do
    avg = avg + pass_times[i]
end
avg = avg / num_passes

local std_div = 0
for i = 1, num_passes do
    local t = pass_times[i]
    local dt = t - avg
    std_div = std_div + (dt * dt)
end
std_div = math.sqrt(std_div / num_passes)

print(math.floor(avg))
print(math.floor(std_div))

os.exit(0)
